Avoid using deprecated Buffer API

lpinca opened the pull request:
Usage of `Buffer` constructor has been deprecated due to usability
issues that can potentially lead to accidental security issues.

Refs: https://nodejs.org/api/deprecations.html#deprecations_dep0005_buffer_constructor
Tracking: https://github.com/nodejs/node/issues/19079

Tom32i commented:
Hey, so where are we on this subject? lpinca's PR looks fine.
Can I help by running tests or in any other way?

lpinca commented:
@Tom32i do you have a test case? By looking at the diff I can't spot any mistake.

Tom32i commented:
I'll try to write a quick runnable test at lunch break so you can run it yourself.

lpinca commented:
Thank you.

lpinca commented:
@Tom32i ping

lpinca commented:
```
$ node server.js 
Listening on localhost:8080
----------
Sending TypedArray Uint8Array [ 255 ]
Sending Buffer <Buffer ff> 1
Sending ArrayBuffer ArrayBuffer { byteLength: 1 } 1
Buffer and ArrayBuffer have same byte length? true
----------
receiving Buffer <Buffer ff> 1
receiving ArrayBuffer ArrayBuffer { byteLength: 1 } 1
Buffer and ArrayBuffer have same byte length? true
^C
luigi@macbook:test-websocket-driver (master)$ git checkout original-ws-driver 
Branch 'original-ws-driver' set up to track remote branch 'original-ws-driver' from 'origin'.
Switched to a new branch 'original-ws-driver'
luigi@macbook:test-websocket-driver (original-ws-driver)$ npm ci
npm WARN prepare removing existing node_modules/ before installation
added 3 packages in 1.214s
luigi@macbook:test-websocket-driver (original-ws-driver)$ node server.js 
Listening on localhost:8080
----------
Sending TypedArray Uint8Array [ 255 ]
Sending Buffer <Buffer ff> 1
Sending ArrayBuffer ArrayBuffer { byteLength: 1 } 1
Buffer and ArrayBuffer have same byte length? true
----------
receiving Buffer <Buffer ff> 1
receiving ArrayBuffer ArrayBuffer { byteLength: 1 } 1
Buffer and ArrayBuffer have same byte length? true
```

This is what I get on Node.js 11.4.0, macOS

lpinca commented:
Also

```
$ node
> var buf = Buffer.from([255])
undefined
> buf.byteLength
1
> buf.buffer.byteLength
8192
>
```

That's the Node.js Buffer work.

lpinca commented:
Yes will do, anyway to convert a `Buffer` to an `ArrayBuffer` you need something like this:

```js
function toArrayBuffer(buf) {
  return buf.buffer.slice(buf.byteOffset, buf.byteOffset + buf.byteLength);
}
```

Tom32i commented:
No, the doc just says `Buffer.buffer`.
And if the Buffer was constructed correctly `Buffer.from(arrayBuffer)` it works perfectly.

```
node 
Buffer.from(new Uint8Array([255]).buffer).buffer
```

lpinca commented:
Yes, I'm talking about `Buffer` to `ArrayBuffer`. https://nodejs.org/api/buffer.html#buffer_buf_byteoffset

lpinca commented:
> No, the documentation recommends Buffer.buffer.
> And if the Buffer was constructed correctly Buffer.from(arrayBuffer) it works perfectly.
>
> ```
> node 
> > Buffer.from(new ArrayBuffer(1)).buffer
> ArrayBuffer { byteLength: 1 }
> ```

This is correct when a `Buffer` is created from an `ArrayBuffer` but see this

```
$ node
> Buffer.from([255,255,255,255]).buffer
ArrayBuffer { byteLength: 8192 }
> Buffer.from('hello').buffer
ArrayBuffer { byteLength: 8192 }
```

jcoglan commented:
Thanks for doing this work @lpinca. Is there a reason you've used `allocUnsafe()` rather than `alloc()`?

lpinca commented:
It's faster as the underlying memory is not initialised.

jcoglan commented:
How much of an impact does it have? Given these APIs were introduced to make it harder for apps to accidentally leak memory, I'd prefer to use the safe version unless there's a very strong reason not to.

lpinca commented:
Well unless there is bug where the buffer length is incorrectly calculated, it is safe as the content is overwritten with the frame data.

The impact should be benchmarked but I think it's not huge.

jcoglan commented:
Cool, that makes sense :)
